### ISSUE TITLE
segfault free () from /usr/lib/x86_64-linux-gnu/libjemalloc.s

### ISSUE THREAD

**Issue by acrosson** [pytorch/pytorch#3111]
I'm getting a segfault a few thousand iterations into using a model for inference. I'm serving the model in a simple flask app. How should I resolve?

Specs
```
Ubuntu 16.04 on AWS g3.8xlarge
Pytorch (latest) installed through conda
python 3.5
cuda 8.0
```

The original error was ```Error in `python': double free or corruption (!prev)```
I've tried resolving it by changing LD_PRELOAD as outlined https://github.com/google/seq2seq/issues/119
```
export LD_PRELOAD="/usr/lib/x86_64-linux-gnu/libjemalloc.so.1"
```
and https://github.com/tensorflow/tensorflow/issues/6968
```
export LD_PRELOAD="/usr/lib/libtcmalloc_minimal.so.4"
```

But these haven't helped. Currently getting
Error 
```
Thread 22951 "python" received signal SIGSEGV, Segmentation fault.
0x00007ffff7ba42d7 in free () from /usr/lib/x86_64-linux-gnu/libjemalloc.s
```

Backtrace from gdb (few hundred lines)
```
#0  0x00007ffff7ba42d7 in free ()
   from /usr/lib/x86_64-linux-gnu/libjemalloc.so.1
#1  0x00007fff8ec55c73 in torch::autograd::VariableVersion::cleanup (
    this=<optimized out>, this=<optimized out>)
   from /home/ubuntu/anaconda3/envs/py3/lib/python3.5/site-packages/torch/_C.cpython-35m-x86_64-linux-gnu.so
#2  torch::autograd::VariableVersion::~VariableVersion (
    this=0x7fff036fd530, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:47
#3  std::default_delete<torch::autograd::VariableVersion>::operator() (
    this=<optimized out>, __ptr=0x7fff036fd530)
    at torch/csrc/autograd/python_function.cpp:67
#4  std::unique_ptr<torch::autograd::VariableVersion, std::default_delete<torch::autograd::VariableVersion> >::~unique_ptr (this=0x7fff0bf5af60, 
    __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:184
#5  torch::autograd::Variable::SavedVariable::~SavedVariable (
    this=0x7fff0bf5af30, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:19
#6  0x00007fff8ec74a65 in torch::autograd::ConvBackward::~ConvBackward (
    this=0x7fff0bf5ae18, __in_chrg=<optimized out>)
   from /home/ubuntu/anaconda3/envs/py3/lib/python3.5/site-packages/torch/_C.cpython-35m-x86_64-linux-gnu.so
#7  0x00007fff8ec68b78 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x7fff0bf5ae00)
   from /home/ubuntu/anaconda3/envs/py3/lib/python3.5/site-packages/torch/_C.cpython-35m-x86_64-linux-gnu.so
#8  std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (
    this=0x7fff02fdf108, __in_chrg=<optimized out>)
    at torch/csrc/autograd/functions/batch_normalization.cpp:546
#9  std::__shared_ptr<torch::autograd::Function, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x7fff02fdf100, __in_chrg=<optimized out>)
    at torch/csrc/autograd/functions/batch_normalization.cpp:781
#10 std::shared_ptr<torch::autograd::Function>::~shared_ptr (
    this=0x7fff02fdf100, __in_chrg=<optimized out>)
    at torch/csrc/autograd/functions/batch_normalization.cpp:93
#11 std::pair<std::shared_ptr<torch::autograd::Function>, int>::~pair (
    this=0x7fff02fdf100, __in_chrg=<optimized out>)
    at torch/csrc/autograd/functions/batch_normalization.cpp:96
#12 std::_Destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int> > (__pointer=0x7fff02fdf100)
    at torch/csrc/autograd/functions/batch_normalization.cpp:93
#13 std::_Destroy_aux<false>::__destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int>*> (__last=<optimized out>, 
    __first=0x7fff02fdf100)
    at torch/csrc/autograd/functions/batch_normalization.cpp:103
#14 std::_Destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int>*> (__last=<optimized out>, __first=<optimized out>)
    at torch/csrc/autograd/functions/batch_normalization.cpp:126
#15 std::_Destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int>*, std::pair<std::shared_ptr<torch::autograd::Function>, int> > (
    __last=0x7fff02fdf148, __first=0x7fff02fdf100)
    at torch/csrc/autograd/functions/batch_normalization.cpp:151
#16 std::vector<std::pair<std::shared_ptr<torch::autograd::Function>, int>, std::allocator<std::pair<std::shared_ptr<torch::autograd::Function>, int> > >::~vector (this=0x7fff0369dea8, __in_chrg=<optimized out>)
    at torch/csrc/autograd/functions/batch_normalization.cpp:415
#17 torch::autograd::Function::~Function (this=0x7fff0369de98, 
    __in_chrg=<optimized out>)
    at torch/csrc/autograd/functions/batch_normalization.cpp:63
#18 torch::autograd::BatchNormBackward::~BatchNormBackward (
    this=0x7fff0369de98, __in_chrg=<optimized out>)
    at torch/csrc/autograd/functions/batch_normalization.cpp:28
#19 0x00007fff8ec4f870 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x7fff0369de80)
   from /home/ubuntu/anaconda3/envs/py3/lib/python3.5/site-packages/torch/_C.cpython-35m-x86_64-linux-gnu.so
#20 std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (
    this=0x7fff02dffb28, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:546
#21 std::__shared_ptr<torch::autograd::Function, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x7fff02dffb20, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:781
#22 std::shared_ptr<torch::autograd::Function>::~shared_ptr (
    this=0x7fff02dffb20, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:93
#23 std::pair<std::shared_ptr<torch::autograd::Function>, int>::~pair (
    this=0x7fff02dffb20, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:96
#24 std::_Destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int> > (__pointer=0x7fff02dffb20)
    at torch/csrc/autograd/python_function.cpp:93
#25 std::_Destroy_aux<false>::__destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int>*> (__last=<optimized out>, 
    __first=0x7fff02dffb20)
    at torch/csrc/autograd/python_function.cpp:103
#26 std::_Destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int>*> (__last=<optimized out>, __first=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:126
#27 std::_Destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int>*, std::pair<std::shared_ptr<torch::autograd::Function>, int> > (
    __last=0x7fff02dffb38, __first=0x7fff02dffb20)
    at torch/csrc/autograd/python_function.cpp:151
#28 std::vector<std::pair<std::shared_ptr<torch::autograd::Function>, int>, std::allocator<std::pair<std::shared_ptr<torch::autograd::Function>, int> > >::~vector (this=0x7fffe78b0bf0, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:415
#29 torch::autograd::Function::~Function (this=0x7fffe78b0be0, 
    __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:63
#30 torch::autograd::PyFunction::~PyFunction (this=0x7fffe78b0be0, 
    __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:19
#31 THPFunction_dealloc (self=0x7fffe78b0b88)
    at torch/csrc/autograd/python_function.cpp:287
#32 0x00007ffff777af89 in subtype_dealloc (self=0x7fffe78b0b88)
    at Objects/typeobject.c:1209
#33 0x00007fff8ec555cb in Decref::operator() (this=<optimized out>, 
    p=0x7fffe78b0be0)
   from /home/ubuntu/anaconda3/envs/py3/lib/python3.5/site-packages/torch/_C.cpython-35m-x86_64-linux-gnu.so
#34 std::_Sp_counted_deleter<torch::autograd::PyFunction*, Decref, std::allocator<int>, (__gnu_cxx::_Lock_policy)2>::_M_dispose (
    this=<optimized out>) at torch/csrc/autograd/python_function.cpp:347
#35 0x00007fff8ec4f870 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x7fff02dffe20)
   from /home/ubuntu/anaconda3/envs/py3/lib/python3.5/site-packages/torch/_C.cpython-35m-x86_64-linux-gnu.so
#36 std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (
    this=0x7fff02dffe08, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:546
#37 std::__shared_ptr<torch::autograd::Function, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x7fff02dffe00, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:781
#38 std::shared_ptr<torch::autograd::Function>::~shared_ptr (
    this=0x7fff02dffe00, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:93
#39 std::pair<std::shared_ptr<torch::autograd::Function>, int>::~pair (
    this=0x7fff02dffe00, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:96
#40 std::_Destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int> > (__pointer=0x7fff02dffe00)
    at torch/csrc/autograd/python_function.cpp:93
#41 std::_Destroy_aux<false>::__destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int>*> (__last=<optimized out>, 
    __first=0x7fff02dffe00)
    at torch/csrc/autograd/python_function.cpp:103
#42 std::_Destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int>*> (__last=<optimized out>, __first=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:126
#43 std::_Destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int>*, std::pair<std::shared_ptr<torch::autograd::Function>, int> > (
    __last=0x7fff02dffe18, __first=0x7fff02dffe00)
    at torch/csrc/autograd/python_function.cpp:151
#44 std::vector<std::pair<std::shared_ptr<torch::autograd::Function>, int>, std::allocator<std::pair<std::shared_ptr<torch::autograd::Function>, int> > >::~vector (this=0x7fffe78b0650, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:415
#45 torch::autograd::Function::~Function (this=0x7fffe78b0640, 
    __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:63
#46 torch::autograd::PyFunction::~PyFunction (this=0x7fffe78b0640, 
    __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:19
#47 THPFunction_dealloc (self=0x7fffe78b05e8)
    at torch/csrc/autograd/python_function.cpp:287
#48 0x00007ffff777af89 in subtype_dealloc (self=0x7fffe78b05e8)
    at Objects/typeobject.c:1209
#49 0x00007fff8ec555cb in Decref::operator() (this=<optimized out>, 
    p=0x7fffe78b0640)
   from /home/ubuntu/anaconda3/envs/py3/lib/python3.5/site-packages/torch/_C.cpython-35m-x86_64-linux-gnu.so
#50 std::_Sp_counted_deleter<torch::autograd::PyFunction*, Decref, std::allocator<int>, (__gnu_cxx::_Lock_policy)2>::_M_dispose (
    this=<optimized out>) at torch/csrc/autograd/python_function.cpp:347
#51 0x00007fff8ec4f870 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x7fff02e00040)
   from /home/ubuntu/anaconda3/envs/py3/lib/python3.5/site-packages/torch/_C.cpython-35m-x86_64-linux-gnu.so
#52 std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (
    this=0x7fff036e9e70, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:546
#53 std::__shared_ptr<torch::autograd::Function, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x7fff036e9e68, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:781
#54 std::shared_ptr<torch::autograd::Function>::~shared_ptr (
    this=0x7fff036e9e68, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:93
#55 std::pair<std::shared_ptr<torch::autograd::Function>, int>::~pair (
    this=0x7fff036e9e68, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:96
#56 std::_Destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int> > (__pointer=0x7fff036e9e68)
    at torch/csrc/autograd/python_function.cpp:93
#57 std::_Destroy_aux<false>::__destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int>*> (__last=<optimized out>, 
    __first=0x7fff036e9e68)
    at torch/csrc/autograd/python_function.cpp:103
#58 std::_Destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int>*> (__last=<optimized out>, __first=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:126
#59 std::_Destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int>*, std::pair<std::shared_ptr<torch::autograd::Function>, int> > (
    __last=0x7fff036e9e80, __first=0x7fff036e9e50)
    at torch/csrc/autograd/python_function.cpp:151
#60 std::vector<std::pair<std::shared_ptr<torch::autograd::Function>, int>, std::allocator<std::pair<std::shared_ptr<torch::autograd::Function>, int> > >::~vector (this=0x7fffef683830, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:415
#61 torch::autograd::Function::~Function (this=0x7fffef683820, 
    __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:63
#62 torch::autograd::PyFunction::~PyFunction (this=0x7fffef683820, 
    __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:19
#63 THPFunction_dealloc (self=0x7fffef6837c8)
    at torch/csrc/autograd/python_function.cpp:287
#64 0x00007ffff777af89 in subtype_dealloc (self=0x7fffef6837c8)
    at Objects/typeobject.c:1209
#65 0x00007fff8ec555cb in Decref::operator() (this=<optimized out>, 
    p=0x7fffef683820)
   from /home/ubuntu/anaconda3/envs/py3/lib/python3.5/site-packages/torch/_C.cpython-35m-x86_64-linux-gnu.so
#66 std::_Sp_counted_deleter<torch::autograd::PyFunction*, Decref, std::allocator<int>, (__gnu_cxx::_Lock_policy)2>::_M_dispose (
    this=<optimized out>) at torch/csrc/autograd/python_function.cpp:347
#67 0x00007fff8ec4f870 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x7fff02e00720)
   from /home/ubuntu/anaconda3/envs/py3/lib/python3.5/site-packages/torch/_C.cpython-35m-x86_64-linux-gnu.so
#68 std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (
    this=0x7fff02e004c8, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:546
#69 std::__shared_ptr<torch::autograd::Function, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x7fff02e004c0, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:781
#70 std::shared_ptr<torch::autograd::Function>::~shared_ptr (
    this=0x7fff02e004c0, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:93
#71 std::pair<std::shared_ptr<torch::autograd::Function>, int>::~pair (
    this=0x7fff02e004c0, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:96
#72 std::_Destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int> > (__pointer=0x7fff02e004c0)
    at torch/csrc/autograd/python_function.cpp:93
#73 std::_Destroy_aux<false>::__destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int>*> (__last=<optimized out>, 
    __first=0x7fff02e004c0)
    at torch/csrc/autograd/python_function.cpp:103
#74 std::_Destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int>*> (__last=<optimized out>, __first=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:126
#75 std::_Destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int>*, std::pair<std::shared_ptr<torch::autograd::Function>, int> > (
    __last=0x7fff02e004d8, __first=0x7fff02e004c0)
    at torch/csrc/autograd/python_function.cpp:151
#76 std::vector<std::pair<std::shared_ptr<torch::autograd::Function>, int>, std::allocator<std::pair<std::shared_ptr<torch::autograd::Function>, int> > >::~vector (this=0x7fffef683bf0, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:415
#77 torch::autograd::Function::~Function (this=0x7fffef683be0, 
    __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:63
#78 torch::autograd::PyFunction::~PyFunction (this=0x7fffef683be0, 
    __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:19
#79 THPFunction_dealloc (self=0x7fffef683b88)
    at torch/csrc/autograd/python_function.cpp:287
#80 0x00007ffff777af89 in subtype_dealloc (self=0x7fffef683b88)
    at Objects/typeobject.c:1209
#81 0x00007fff8ec555cb in Decref::operator() (this=<optimized out>, 
    p=0x7fffef683be0)
   from /home/ubuntu/anaconda3/envs/py3/lib/python3.5/site-packages/torch/_C.cpython-35m-x86_64-linux-gnu.so
#82 std::_Sp_counted_deleter<torch::autograd::PyFunction*, Decref, std::allocator<int>, (__gnu_cxx::_Lock_policy)2>::_M_dispose (
    this=<optimized out>) at torch/csrc/autograd/python_function.cpp:347
#83 0x00007fff8ec4f870 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x7fff02e007c0)
   from /home/ubuntu/anaconda3/envs/py3/lib/python3.5/site-packages/torch/_C.cpython-35m-x86_64-linux-gnu.so
#84 std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (
    this=0x7fff037070f0, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:546
#85 std::__shared_ptr<torch::autograd::Function, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x7fff037070e8, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:781
#86 std::shared_ptr<torch::autograd::Function>::~shared_ptr (
    this=0x7fff037070e8, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:93
#87 std::pair<std::shared_ptr<torch::autograd::Function>, int>::~pair (
    this=0x7fff037070e8, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:96
#88 std::_Destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int> > (__pointer=0x7fff037070e8)
    at torch/csrc/autograd/python_function.cpp:93
#89 std::_Destroy_aux<false>::__destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int>*> (__last=<optimized out>, 
    __first=0x7fff037070e8)
    at torch/csrc/autograd/python_function.cpp:103
#90 std::_Destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int>*> (__last=<optimized out>, __first=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:126
#91 std::_Destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int>*, std::pair<std::shared_ptr<torch::autograd::Function>, int> > (
    __last=0x7fff03707100, __first=0x7fff037070d0)
    at torch/csrc/autograd/python_function.cpp:151
#92 std::vector<std::pair<std::shared_ptr<torch::autograd::Function>, int>, std::allocator<std::pair<std::shared_ptr<torch::autograd::Function>, int> > >::~vector (this=0x7fffef683dd0, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:415
#93 torch::autograd::Function::~Function (this=0x7fffef683dc0, 
    __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:63
#94 torch::autograd::PyFunction::~PyFunction (this=0x7fffef683dc0, 
    __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:19
#95 THPFunction_dealloc (self=0x7fffef683d68)
    at torch/csrc/autograd/python_function.cpp:287
#96 0x00007ffff777af89 in subtype_dealloc (self=0x7fffef683d68)
    at Objects/typeobject.c:1209
#97 0x00007fff8ec555cb in Decref::operator() (this=<optimized out>, 
    p=0x7fffef683dc0)
   from /home/ubuntu/anaconda3/envs/py3/lib/python3.5/site-packages/torch/_C.cpython-35m-x86_64-linux-gnu.so
#98 std::_Sp_counted_deleter<torch::autograd::PyFunction*, Decref, std::allocator<int>, (__gnu_cxx::_Lock_policy)2>::_M_dispose (
    this=<optimized out>) at torch/csrc/autograd/python_function.cpp:347
#99 0x00007fff8ec4f870 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x7fff02e00ec0)
   from /home/ubuntu/anaconda3/envs/py3/lib/python3.5/site-packages/torch/_C.cpython-35m-x86_64-linux-gnu.so
#100 std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (
    this=0x7fff02e00c68, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:546
#101 std::__shared_ptr<torch::autograd::Function, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x7fff02e00c60, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:781
#102 std::shared_ptr<torch::autograd::Function>::~shared_ptr (
    this=0x7fff02e00c60, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:93
#103 std::pair<std::shared_ptr<torch::autograd::Function>, int>::~pair (
    this=0x7fff02e00c60, __in_chrg=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:96
#104 std::_Destroy<std::pair<std::shared_ptr<torch::autograd::Function>, int> > (__pointer=0x7fff02e00c60)
    at torch/csrc/autograd/python_function.cpp:93
#105
```
I have observed the memory via htop and nvidia-smi manually every few hundred iterations and didn't notice any significant uptick in memory growth over time.

**Comment by soumith** [comment]
thanks a lot for giving such a detailed report, it really helps.

@colesbury do you have any ideas where this might be popping up from?

**Comment by apaszke** [comment]
Is there any chance we could see the model, or at least a simplified example that could cause the crash? Do you operate on the same autograd graph in multiple threads?

**Comment by colesbury** [comment]
To expand on @apaszke's answer, the refcounting in `VariableVersion` isn't thread-safe. It seems likely that this is related. We'll fix the refcounting before the next release.

**Comment by acrosson** [comment]
I have a few models that are being ran, one of which is a modified version of ResNet from pytorch code base https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py ... 

I run the flask app with threaded=True
```
import jsonpickle
import numpy as np
import cv2
import json

from flask import Flask, request, Response
from model import Model

app = Flask(__name__)
m = Model()

@app.route('/api/detect', methods=['POST'])
def detect():
    r = request
    rgb_im_len = int(r.headers.get('rgb_im_len'))
    # convert string of image data to uint8
    nparr = np.fromstring(r.data[:rgb_im_len], np.uint8)
    im = cv2.imdecode(nparr, cv2.IMREAD_COLOR) # decode image

    output = m.detect(im)
    response = {'output': output}
    response_pickled = jsonpickle.encode(response)
    return Response(response=response_pickled, status=200, mimetype="application/json")


# start flask app
app.run(host="0.0.0.0", port=5000, threaded=True)
```

**Comment by colesbury** [comment]
@acrosson, this should be resolved in `master`. Can you let us know if it fixes the segfault for you?

**Comment by acrosson** [comment]
@colesbury awesome, ill test it out and circle back

**Comment by acrosson** [comment]
I installed from source, using `cuda V8.0.61`. There now appears to be a memory leak, every iteration increases the GPU memory, until eventually throwing 

```
RuntimeError: cuda runtime error (2) : out of memory at /home/ubuntu/pytorch/torch/lib/THC/generic/THCStorage.cu:58
```

**Comment by colesbury** [comment]
@acrosson, thanks for checking it out. Is it the same code as above? If so, I'll try it out on my side.

**Comment by acrosson** [comment]
@colesbury yup, nothing has been modified . thanks!

**Comment by colesbury** [comment]
@acrosson, can you share code for `class Model`? I'd like to see how the image gets wrapped in a Variable and passed to the network.

**Comment by colesbury** [comment]
@acrosson, I wrote up a test script here:
https://github.com/colesbury/issue_3111

I'm not seeing any leak of GPU memory. Does this setup leak for you? What's different from your setup?

**Comment by acrosson** [comment]
@colesbury 
There's two models that are being called in the `Model` class. One is the ResNet and the other is https://github.com/amdegroot/ssd.pytorch.

I ran your code and it works fine - no mem leak. I also modified your repo to include ssd.pytorch and it worked fine. So the memory leak must be an error on my part. Weird that it doesn't happen with the latest released version of pytorch though. 

Let me track it down and then i'll test for the segfault.
